cluster-agent/controllers/argoproj.io/namespace_reconciler.go

```diff
@@ -0,0 +1,335 @@
+	// Get list of Operations from cluster.
+	listOfK8sOperation := v1alpha1.OperationList{}
+	err := client.List(ctx, &listOfK8sOperation)
+	if err != nil {
+		log.Error(err, "Unable to fetch list of k8s Operation from cluster.")
+		return
+	}
+
+	for _, k8sOperation := range listOfK8sOperation.Items {
+
+		// Skip if Operation was not created by Namespace Reconciler.
+		if k8sOperation.Annotations[appEventLoop.IdentifierKey] != appEventLoop.IdentifierValue {
+			continue
+		}
+
+		// Fetch corresponding DB entry
+		dbOperation := db.Operation{
+			Operation_id: k8sOperation.Spec.OperationID,
+		}
+		if err := dbq.GetOperationById(ctx, &dbOperation); err != nil {
+			continue
+		}
+
+		if dbOperation.State != db.OperationState_Completed && dbOperation.State != db.OperationState_Failed {
+			log.V(sharedutil.LogLevel_Debug).Info("K8s Operation is not ready for cleanup : " + string(k8sOperation.UID) + " DbOperation: " + string(k8sOperation.Spec.OperationID))
+			continue
+		}
+
+		log.Info("Deleting Operation created by Namespace Reconciler." + string(k8sOperation.UID))
+
+		// Delete the k8s operation now.
+		if err := appEventLoop.CleanupOperation(ctx, dbOperation, k8sOperation, cache.GetGitOpsEngineSingleInstanceNamespace(), dbq, client, log); err != nil {
+			log.Error(err, "Unable to Delete k8s Operation"+string(k8sOperation.UID)+" for DbOperation: "+string(k8sOperation.Spec.OperationID))
+		} else {
+			log.Info("Deleted k8s Operation: " + string(k8sOperation.UID) + " for DbOperation: " + string(k8sOperation.Spec.OperationID))
+		}
+	}
+	log.V(sharedutil.LogLevel_Debug).Info("Cleaned all Operations created by Namespace Reconciler.")
+}
+
+func deleteOrphanedApplications(argoApplications []appv1.Application, processedApplicationIds map[string]string, ctx context.Context, client client.Client, log logr.Logger) []appv1.Application {
+	log.V(sharedutil.LogLevel_Debug).Info("Looking for any orphaned ArgoCD application to delete.")
+
+	// Iterate through all application IDs collected from DB.
+	for id := range processedApplicationIds {
+		// Iterate through ArgoCD all and remove corresponding application from list.
+		for i, application := range argoApplications {
+			if application.Labels["databaseID"] == id {
+				argoApplications = dropProcessedApplicationId(argoApplications, i)
+			}
+		}
+	}
+
+	log.Info(fmt.Sprintf("Found %d orphaned ArgoCD application to be deleted.", len(argoApplications)))
+	// Delete all applications which are not having entry in DB.
+	deleteApplication(ctx, client, argoApplications, log)
+
+	return argoApplications
+}
+
+func dropProcessedApplicationId(argoApplications []appv1.Application, i int) []appv1.Application {
+	argoApplications[i] = argoApplications[len(argoApplications)-1]
+	return argoApplications[:len(argoApplications)-1]
+}
+
+func deleteApplication(ctx context.Context, client client.Client, argoApplications []appv1.Application, log logr.Logger) {
+	for _, application := range argoApplications {
```

Comment:
I wonder if we could directly delete the application in the previous function without looping through them again?

Reply:
Can you elaborate a bit more, please?
In previous function I am removing all those applications from list, which are having entry in DB. As result of this all orphaned applications remaining in list, so in this function I am going through that list in deleting applications from ArgoCD one by one.

Reply:
Same as [above](https://github.com/redhat-appstudio/managed-gitops/pull/169#discussion_r919839837). If we delete the orphaned applications from the DB in the previous function itself we don't have to call  `deleteApplication()` and loop again.